### PR TITLE
[Enhancement][EnemyAI] Add support for simulated damage calculations and "Search for KO" move filtering

### PR DESCRIPTION
**Pull request by innerthunder**
## What are the changes the user will see?
The AI won't use random status moves instead of going for a KO anymore.

## Why am I making these changes?
In the mainline games and many popular ROM Hacks, the enemy AI uses damage calculations to limit their move selection to moves that can KO a player Pokemon when possible. Until the recently added support for simulated ability calls, it wasn't possible for the enemy AI in PokeRogue to do so without introducing major side effects to the game state. However, with that support now added, it's now possible to decouple damage modifiers from the central `apply` function and silently calculate damage to implement this behavior.

## What are the changes from a developer perspective?
`field/pokemon`:
- All parts of the damage calculation in `Pokemon#apply` (except for critical hit resolution) are now moved to the function `Pokemon#getAttackDamage`, which takes several parameters:
  - `ignoreAbility`: If `true`, ignores the defending Pokemon's abilities during the calculation. Currently used for damage calculations where the player Pokemon's ability is unknown to the enemy.
  - `ignoreSourceAbility`: If `true`, ignores the attacking Pokemon's abilities during the calculation. Currently unused, but can be useful for when the enemy is assessing its defensive matchup against a player Pokemon whose ability is "unknown."
  - `simulated`: If `true`, suppresses all effects that would trigger messages or ability popups, or otherwise change the game state. This is only set to `false` when this function is called in `apply`.
- A few bugs were fixed in the move to `getAttackDamage`, the most notable being that the STAB damage multiplier now correctly stacks the effects of the attacker's type(s) and Tera type. All damage multipliers internal to the calculation are also documented with comments.
- If one or more of the enemy's moves KO at least one player Pokemon, `EnemyPokemon#getNextMove` will only evaluate the moves that can KO for selection.

`test/enemy_command.test`: New probabilistic test for `EnemyPokemon#getNextMove`. Currently only checks that status moves are never used when a KO is possible.

`test/battle/damage_calculation.test`: Added tests for `Pokemon.getAttackDamage`
- Attacks should deal 1 damage at minimum
- Fixed-damage moves aren't affected by damage multipliers (specifically testing against type multipliers and Multiscale)
- One-hit KO moves aren't affected by damage multipliers (same as above)

#### Move Choice Analysis: Before and After "Search for KO"

The test in `enemy_command.test` simulates the following battle scenario:
- The player Pokemon is a Level 1 Magikarp
- The enemy Pokemon is a Level 100 Eternatus with the moves Eternabeam, Sludge Bomb, Cosmic Power, and Dragon Dance
- Both Pokemon are given useless abilities (Ball Fetch, more specifically)

Basically, this means that any attack from the enemy Eternatus will KO the player's Magikarp.

This test calls `getNextMove` 300 times in a single turn to estimate the probability of each move being selected. Below is the result of this test when the "Search for KO" filtering is not implemented. As you can see, even though the enemy can KO the player with either Eternabeam or Sludge Bomb, the other moves have a fairly significant chance of being selected.

EnemyAIPreKOSearch

On the other hand, after "Search for KO" filtering, the enemy is guaranteed to select a move that KOs the Magikarp.

EnemyAIPostKOSearch

Note that this test is on the `SMART_RANDOM` enemy AI type, which is widely used for Wild Pokemon. The `SMART` AI type, which is used for Trainer Pokemon, employs a stricter score-based filter in its move selection, so differences in behavior before and after "Search for KO" are likely less noticeable.

## How to test the changes?
`npm run test enemy_command`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [?] Are the changes visual?
  - [?] Have I provided screenshots/videos of the changes?

### CODE REVIEW

**Review by frutescens — APPROVED** [review]

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -3943,6 +4020,31 @@ export class EnemyPokemon extends Pokemon {
         return { move: moveId, targets: this.getNextTargets(moveId) };
       case AiType.SMART_RANDOM:
       case AiType.SMART:
+        /**
+         * Search this Pokemon's move pool for moves that will KO an opposing target.
+         * If there are any moves that can KO an opponent (i.e. a player Pokemon),
+         * those moves are the only ones considered for selection on this turn.
+         */
+        const killMoves = movePool.filter(pkmnMove => {
```
[the remark this answers is not in the document]
Comment (innerthunder) [reply src/field/pokemon.ts]: Renamed to "koMoves"

**Review by frutescens — APPROVED** [review]

File: src/field/pokemon.ts [hunk 2]
```diff
@@ -3922,7 +3999,7 @@ export class EnemyPokemon extends Pokemon {
     }
 
     // Filter out any moves this Pokemon cannot use
-    const movePool = this.getMoveset().filter(m => m?.isUsable(this));
+    let movePool = this.getMoveset().filter(m => m?.isUsable(this));
```
Comment (DayKev) [inline src/field/pokemon.ts:4002]: Why remove the `const` if this doesn't get redefined later?

File: src/field/pokemon.ts [hunk 3]
```diff
@@ -3943,6 +4020,31 @@ export class EnemyPokemon extends Pokemon {
...
+          if (!pkmnMove) {
+            return false;
+          }
+          const move = pkmnMove.getMove()!;
+          const fieldPokemon = this.scene.getField();
+          const moveTargets = getMoveTargets(this, move.id).targets
+            .map(ind => fieldPokemon[ind])
+            .filter(p => this.isPlayer() !== p.isPlayer());
+          // Only considers critical hits for crit-only moves or when this Pokemon is under the effect of Laser Focus
+          const isCritical = move.hasAttr(CritOnlyAttr) || !!this.getTag(BattlerTagType.ALWAYS_CRIT);
```
Comment (DayKev) [inline src/field/pokemon.ts:4038]:
```suggestion
          const isCritical = (move.hasAttr(CritOnlyAttr) || this.getTag(BattlerTagType.ALWAYS_CRIT));
```
I don't think you need to do `!!` here.

File: src/test/battle/damage_calculation.test.ts [hunk 4]
```diff
@@ -25,46 +24,95 @@ describe("Round Down and Minimun 1 test in Damage Calculation", () => {
...
-    game.override.battleType("single");
-    game.override.startingLevel(10);
+    game.override
+      .battleType("single")
+      .enemySpecies(Species.SNORLAX)
+      .enemyAbility(Abilities.BALL_FETCH)
+      .enemyMoveset(SPLASH_ONLY)
+      .startingLevel(100)
+      .enemyLevel(100)
+      .disableCrits();
```
Comment (DayKev) [inline src/test/battle/damage_calculation.test.ts:34]:
```suggestion
      .disableCrits()
      .moveset([Moves.TACKLE, Moves.DRAGON_RAGE, Moves.FISSURE, Moves.JUMP_KICK]);
```

File: src/test/battle/damage_calculation.test.ts [hunk 5]
```diff
@@ -25,46 +24,95 @@ describe("Round Down and Minimun 1 test in Damage Calculation", () => {
...
+  });
+
+  it("Attacks deal 1 damage at minimum", async () => {
+    game.override
+      .startingLevel(1)
+      .enemySpecies(Species.AGGRON)
+      .moveset([Moves.TACKLE]);
```
Comment (DayKev) [inline src/test/battle/damage_calculation.test.ts:41]:
```suggestion
      .enemySpecies(Species.AGGRON);
```

File: src/test/battle/damage_calculation.test.ts [hunk 6]
```diff
@@ -25,46 +24,95 @@ describe("Round Down and Minimun 1 test in Damage Calculation", () => {
...
+
+    // Lvl 1 0 Atk Magikarp Tackle vs. 0 HP / 0 Def Aggron: 1-1 (0.3 - 0.3%) -- possibly the worst move ever
+    expect(aggron.hp).toBe(aggron.getMaxHp() - 1);
+  });
+
+  it("Fixed-damage moves ignore damage multipliers", async () => {
+    game.override
+      .enemySpecies(Species.DRAGONITE)
+      .enemyAbility(Abilities.MULTISCALE)
+      .moveset([Moves.DRAGON_RAGE]);
```
Comment (DayKev) [inline src/test/battle/damage_calculation.test.ts:59]:
```suggestion
      .enemyAbility(Abilities.MULTISCALE);
```

File: src/test/battle/damage_calculation.test.ts [hunk 7]
```diff
@@ -25,46 +24,95 @@ describe("Round Down and Minimun 1 test in Damage Calculation", () => {
...
+    const dragonite = game.scene.getEnemyPokemon()!;
+
+    expect(dragonite.getAttackDamage(magikarp, allMoves[Moves.DRAGON_RAGE]).damage).toBe(40);
+  });
+
+  it("One-hit KO moves ignore damage multipliers", async () => {
+    game.override
+      .enemySpecies(Species.AGGRON)
+      .enemyAbility(Abilities.MULTISCALE)
+      .moveset([Moves.FISSURE]);
```
Comment (DayKev) [inline src/test/battle/damage_calculation.test.ts:73]:
```suggestion
      .enemyAbility(Abilities.MULTISCALE);
```

File: src/test/battle/damage_calculation.test.ts [hunk 8]
```diff
@@ -25,46 +24,95 @@ describe("Round Down and Minimun 1 test in Damage Calculation", () => {
...
 
   it("When the user fails to use Jump Kick with Wonder Guard ability, the damage should be 1.", async () => {
-    game.override.enemySpecies(Species.GASTLY);
-    game.override.enemyMoveset(SPLASH_ONLY);
-    game.override.starterSpecies(Species.SHEDINJA);
-    game.override.moveset([Moves.JUMP_KICK]);
-    game.override.ability(Abilities.WONDER_GUARD);
+    game.override
+      .enemySpecies(Species.GASTLY)
+      .moveset([Moves.JUMP_KICK])
```
Comment (DayKev) [inline src/test/battle/damage_calculation.test.ts:86]:
```suggestion
```

File: src/field/pokemon.ts [hunk 9]
```diff
@@ -2021,301 +2033,366 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
-        }
+    const isTypeImmune = (typeMultiplier * arenaAttackTypeMultiplier.value) === 0;
 
-        /** Combined damage multiplier from field effects such as weather, terrain, etc. */
-        const arenaAttackTypeMultiplier = new Utils.NumberHolder(this.scene.arena.getAttackTypeMultiplier(moveType, source.isGrounded()));
-        applyMoveAttrs(IgnoreWeatherTypeDebuffAttr, source, this, move, arenaAttackTypeMultiplier);
+    if (cancelled.value || isTypeImmune) {
+      return {
+        cancelled: cancelled.value,
+        result: move.id === Moves.SHEER_COLD ? HitResult.IMMUNE : HitResult.NO_EFFECT,
```
[the remark this answers is not in the document]
Comment (innerthunder) [reply src/field/pokemon.ts:2053]: The original code only uses `HitResult.IMMUNE` for Sheer Cold, as apparently it has a different message when used against Ice types (to no effect)

File: src/field/pokemon.ts (hunk 3 above)
Comment (innerthunder) [reply src/field/pokemon.ts:4038]: I tried removing it here, TS got mad at me

File: src/field/pokemon.ts (hunk 2 above)
Comment (innerthunder) [reply src/field/pokemon.ts:4002]: It can be redefined later, see line 4044

Comment (DayKev) [reply src/field/pokemon.ts:4143]: Oh I missed that line, nvm lol.

File: src/field/pokemon.ts (hunk 3 above)
Comment (DayKev) [reply src/field/pokemon.ts:4179]: Hm, alright. Not a big deal then.

File: src/field/pokemon.ts [hunk 10]
```diff
@@ -2021,301 +2033,366 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
+
+    /** The damage multiplier when the given move critically hits */
+    const criticalMultiplier = new Utils.NumberHolder(isCritical ? 1.5 : 1);
+    applyAbAttrs(MultCritAbAttr, source, null, simulated, criticalMultiplier);
+
+    /**
+     * A multiplier for random damage spread in the range [0.85, 1]
+     * This is always 1 for simulated calls.
+     */
+    const randomMultiplier = simulated ? 1 : ((this.scene.randBattleSeedInt(16) + 85) / 100);
```
[the remark this answers is not in the document]
Comment (innerthunder) [reply src/field/pokemon.ts]: So I changed this, but I also had to change `GameManager`'s rSI prototype:
```ts
BattleScene.prototype.randBattleSeedInt = (range, min) => (min ?? 0) + range - 1;
```
It used to not account for `min`, so for example `Pokemon.randSeedIntRange(85, 100)` would return 15 in tests before the change.

**Review by damocleas — DISMISSED** [review]
aha.

File: src/field/pokemon.ts [hunk 11]
```diff
@@ -2156,21 +2167,29 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
-  * @returns {HitResult} The result of the attack
-  */
-  apply(source: Pokemon, move: Move): HitResult {
-    let result: HitResult;
+   * Calculates the damage of an attack made by another Pokemon against this Pokemon
+   * @param source {@linkcode Pokemon} the attacking Pokemon
+   * @param move {@linkcode Pokemon} the move used in the attack
+   * @param ignoreAbility `boolean` if `true`, ignores this Pokemon's defensive ability effects
+   * @param isCritical `boolean` if `true`, calculates damage for a critical hit.
+   * @param simulated `boolean` if `true`, suppresses changes to game state during the calculation.
```
Comment (DayKev) [inline src/field/pokemon.ts:2181]:
```suggestion
   * @param ignoreAbility If `true`, ignores this Pokemon's defensive ability effects
   * @param isCritical If `true`, calculates damage for a critical hit.
   * @param simulated If `true`, suppresses changes to game state during the calculation.
```

File: src/field/pokemon.ts [hunk 12]
```diff
@@ -2156,21 +2167,29 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
+   * @returns a {@linkcode DamageCalculationResult} object with four fields:
+   * - `move`: {@linkcode Moves} the identifier for the attacking move.
+   * - `cancelled`: `boolean` whether the move was cancelled by another effect.
```
Comment (DayKev) [inline src/field/pokemon.ts:2184]:
```suggestion
   * - `cancelled`: `true` if the move was cancelled by another effect.
```

File: src/field/pokemon.ts [hunk 13]
```diff
@@ -2184,279 +2203,344 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
-        if (!fixedDamage.value && !isOneHitKo) {
-          if (!source.isPlayer()) {
-            this.scene.applyModifiers(EnemyDamageBoosterModifier, false, damage);
-          }
-          if (!this.isPlayer()) {
-            this.scene.applyModifiers(EnemyDamageReducerModifier, false, damage);
-          }
+    /**
+     * For each {@link HitsTagAttr} the move has, doubles the damage of the move if:
+     * The target has a {@link BattlerTagType} that this move interacts with
```
Comment (DayKev) [inline src/field/pokemon.ts:2344]:
```suggestion
     * For each {@linkcode HitsTagAttr} the move has, doubles the damage of the move if:
     * The target has a {@linkcode BattlerTagType} that this move interacts with
```

File: src/field/pokemon.ts [hunk 14]
```diff
@@ -2156,21 +2167,29 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
-  apply(source: Pokemon, move: Move): HitResult {
-    let result: HitResult;
+   * Calculates the damage of an attack made by another Pokemon against this Pokemon
+   * @param source {@linkcode Pokemon} the attacking Pokemon
+   * @param move {@linkcode Pokemon} the move used in the attack
+   * @param ignoreAbility `boolean` if `true`, ignores this Pokemon's defensive ability effects
+   * @param isCritical `boolean` if `true`, calculates damage for a critical hit.
+   * @param simulated `boolean` if `true`, suppresses changes to game state during the calculation.
+   * @returns a {@linkcode DamageCalculationResult} object with four fields:
+   * - `move`: {@linkcode Moves} the identifier for the attacking move.
```
Comment (DayKev) [inline src/field/pokemon.ts:2183]: It doesn't actually return `move` (any more?) though.

File: src/field/pokemon.ts [hunk 15]
```diff
@@ -2184,279 +2203,344 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
+      damage: damage.value
+    };
+  }
+
+  /**
+  * Apply the results of a move to this pokemon
+  * @param {Pokemon} source The pokemon using the move
+  * @param {PokemonMove} battlerMove The move being used
+  * @returns {HitResult} The result of the attack
+  */
```
Comment (DayKev) [inline src/field/pokemon.ts:2427]:
```suggestion
  /**
   * Applies the results of a move to this pokemon
   * @param source The {@linkcode Pokemon} using the move
   * @param move The {@linkcode Move} being used
   * @returns The {@linkcode HitResult} of the attack
   */
```

**Review by DayKev — DISMISSED** [review]

**Comment by DayKev** [comment]
If no one else finds changes that need to be made, make sure to make a quick follow-up PR to fix the tsdocs.

File: src/field/pokemon.ts (hunk 14 above)
Comment (innerthunder) [reply src/field/pokemon.ts:2183]:
```suggestion
   * @returns a {@linkcode DamageCalculationResult} object with three fields:
```
yeah, this was from an old version of the code

**Review by DayKev — DISMISSED** [review]

**Review by damocleas — APPROVED** [review]

**Review by damocleas — DISMISSED** [review]

**Review by DayKev — APPROVED** [review]

File: src/field/pokemon.ts [hunk 16]
```diff
@@ -4779,6 +4885,12 @@ export enum HitResult {
 
 export type DamageResult = HitResult.EFFECTIVE | HitResult.SUPER_EFFECTIVE | HitResult.NOT_VERY_EFFECTIVE | HitResult.ONE_HIT_KO | HitResult.OTHER;
 
+export interface DamageCalculationResult {
+  cancelled: boolean;
+  result: HitResult;
+  damage: number;
+}
```
Comment (xsn34kzx) [inline src/field/pokemon.ts:4892]: For posterity, I'd say throw some docs on here (although, the interface and fields names are pretty self-explanatory).

File: src/field/pokemon.ts [hunk 17]
```diff
@@ -2304,21 +2315,28 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
+   * - `cancelled`: `true` if the move was cancelled by another effect.
+   * - `result`: {@linkcode HitResult} indicates the attack's type effectiveness.
+   * - `damage`: `number` the attack's final damage output.
+   */
+  getAttackDamage(source: Pokemon, move: Move, ignoreAbility: boolean = false, ignoreSourceAbility: boolean = false, isCritical: boolean = false, simulated: boolean = true): DamageCalculationResult {
     const damage = new Utils.NumberHolder(0);
     const defendingSide = this.isPlayer() ? ArenaTagSide.PLAYER : ArenaTagSide.ENEMY;
 
-    const variableCategory = new Utils.NumberHolder(move.category);
+    const variableCategory = new Utils.IntegerHolder(move.category);
```
Comment (xsn34kzx) [inline src/field/pokemon.ts:2339]: Does this have to be an `IntegerHolder`?

**Review by xsn34kzx — COMMENTED** [review]
Had some minor comments, but overall this lgtm!

File: src/field/pokemon.ts (hunk 17 above)
Comment (DayKev) [reply src/field/pokemon.ts:2339]: `IntegerHolder` is literally just `NumberHolder` with a different name, so no it doesn't.

Comment (innerthunder) [reply src/field/pokemon.ts:2339]: Most `VariableCategory` attributes cast this to a `NumberHolder` (except for Shell Side Arm's attribute for some reason), so I can change it back. Not really sure how this changed to `IntegerHolder` in the first place...

**Review by DayKev — APPROVED** [review]

**Review by xsn34kzx — APPROVED** [review]

**Review by damocleas — APPROVED** [review]

**Review by Madmadness65 — APPROVED** [review]

**Pull request merged**